pkg/sources/elasticsearch/api.go

```diff
@@ -0,0 +1,471 @@
+package elasticsearch
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"slices"
+	"strings"
+	"sync"
+	"time"
+
+	es "github.com/elastic/go-elasticsearch/v8"
+	"github.com/elastic/go-elasticsearch/v8/esapi"
+)
+
+const PAGE_SIZE = 10
+
+type IndexStatus int
+
+type FilterParams struct {
+	indexPattern   string
+	queryJSON      string
+	sinceTimestamp string
+}
+
+type PointInTime struct {
+	ID        string `json:"id"`
+	KeepAlive string `json:"keep_alive"`
+}
+
+type SearchRequestBody struct {
+	PIT         PointInTime    `json:"pit"`
+	Sort        []string       `json:"sort"`
+	SearchAfter []int          `json:"search_after,omitempty"`
+	Query       map[string]any `json:"query,omitempty"`
+}
+
+type Document struct {
+	id        string
+	timestamp string
+	message   string
+}
+
+type Index struct {
+	name                   string
+	documentCount          int
+	latestTimestamp        time.Time
+	latestTimestampLastRun time.Time
+	latestDocumentIDs      []string
+	lock                   sync.RWMutex
+}
+
+type Indices struct {
+	indices                 []*Index
+	documentCount           int
+	processedDocumentsCount int
+	filterParams            *FilterParams
+	lock                    sync.RWMutex
+}
+
+type elasticSearchRequest interface {
+	Do(providedCtx context.Context, transport esapi.Transport) (*esapi.Response, error)
+}
+
+func (fp *FilterParams) Query(latestTimestamp time.Time) (map[string]any, error) {
+	range_ := make(map[string]any)
+
+	if fp.queryJSON != "" {
+		err := json.Unmarshal([]byte(fp.queryJSON), &range_)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if !latestTimestamp.IsZero() {
+		gte := make(map[string]string)
+		gte["gte"] = latestTimestamp.Format(time.RFC3339)
+
+		timestamp := make(map[string]map[string]string)
+		timestamp["@timestamp"] = gte
+
+		range_["range"] = timestamp
+	} else if fp.sinceTimestamp != "" {
+		gte := make(map[string]string)
+		gte["gte"] = fp.sinceTimestamp
+
+		timestamp := make(map[string]map[string]string)
+		timestamp["@timestamp"] = gte
+
+		range_["range"] = timestamp
+	}
+
+	query := make(map[string]any)
+	query["query"] = range_
+
+	return query, nil
+}
+
+func NewIndex() *Index {
+	return &Index{}
+}
+
+func (i *Index) DocumentAlreadySeen(document *Document) bool {
+	i.lock.Lock()
+	defer i.lock.Unlock()
```

Comment:
**Suggestion**: rather than locking at the top and deferring the unlock we should just surround contentious access with straight up lock/unlocks rather than defer. Easier to figure out what needs to be protected that way.